src/cz/crcs/ectester/standalone/libs/jni/libressl.c

```diff
@@ -0,0 +1,537 @@
+        for (size_t i = 0; i < ncurves; ++i) {
+            if (strcasecmp(utf_name, OBJ_nid2sn(curves[i].nid)) == 0) {
+                curve = EC_GROUP_new_by_curve_name(curves[i].nid);
+                break;
+            }
+        }
+        (*env)->ReleaseStringUTFChars(env, name, utf_name);
+        if (!curve) {
+            throw_new(env, "java/security/InvalidAlgorithmParameterException", "Curve for given bitsize not found.");
+            return NULL;
+        }
+        jobject result = generate_from_curve(env, curve);
+        EC_GROUP_free(curve);
+        return result;
+    } else {
+        throw_new(env, "java/security/InvalidAlgorithmParameterException", "Curve not found.");
+        return NULL;
+    }
+}
+
+EC_KEY *barray_to_pubkey(JNIEnv *env, const EC_GROUP *curve, jbyteArray pub) {
+    EC_KEY *result = EC_KEY_new();
+    EC_KEY_set_group(result, curve);
+    jsize pub_len = (*env)->GetArrayLength(env, pub);
+    jbyte *pub_data = (*env)->GetByteArrayElements(env, pub, NULL);
+    EC_POINT *pub_point = EC_POINT_new(curve);
+    EC_POINT_oct2point(curve, pub_point, (unsigned char *) pub_data, pub_len, NULL);
+    (*env)->ReleaseByteArrayElements(env, pub, pub_data, JNI_ABORT);
+    EC_KEY_set_public_key(result, pub_point);
+    EC_POINT_free(pub_point);
+    return result;
+}
+
+EC_KEY *barray_to_privkey(JNIEnv *env,  const EC_GROUP *curve, jbyteArray priv) {
+    EC_KEY *result = EC_KEY_new();
+    EC_KEY_set_group(result, curve);
+    jsize priv_len = (*env)->GetArrayLength(env, priv);
+    jbyte *priv_data = (*env)->GetByteArrayElements(env, priv, NULL);
+    BIGNUM *s = BN_bin2bn((unsigned char *) priv_data, priv_len, NULL);
+    (*env)->ReleaseByteArrayElements(env, priv, priv_data, JNI_ABORT);
+    EC_KEY_set_private_key(result, s);
+    BN_free(s);
+    return result;
+}
+
+JNIEXPORT jbyteArray JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeKeyAgreementSpi_00024Libressl_generateSecret___3B_3BLjava_security_spec_ECParameterSpec_2(JNIEnv *env, jobject self, jbyteArray pubkey, jbyteArray privkey, jobject params) {
+    EC_GROUP *curve = create_curve(env, params);
+    if (!curve) {
+        throw_new(env, "java/security/InvalidAlgorithmParameterException", "Curve not found.");
+        return NULL;
+    }
+
+    EC_KEY *pub = barray_to_pubkey(env, curve, pubkey);
+    EC_KEY *priv = barray_to_privkey(env, curve, privkey);
+
+    int field_size = EC_GROUP_get_degree(curve);
+    size_t secret_len = (field_size + 7)/8;
+
+    //TODO: Do more KeyAgreements here, but will have to do the hash-fun manually,
+    //      probably using the ECDH_KDF_X9_62 by wrapping it and dynamically choosing the EVP_MD. from the type string.
+    jbyteArray result = (*env)->NewByteArray(env, secret_len);
+    jbyte *result_data = (*env)->GetByteArrayElements(env, result, NULL);
+
+    native_timing_start();
+    int err = ECDH_compute_key(result_data, secret_len, EC_KEY_get0_public_key(pub), priv, NULL);
```

Comment:
The `get0_public_key` should happen outside of the timed part.

Reply:
Yup, will fix. Thanks for pointing that out.

Reply:
Actually, does it matter? It is [defined](https://github.com/libressl-portable/openbsd/blob/ce8330bc56df642246f659641e068879414d4021/src/lib/libcrypto/ec/ec_key.c#L484-L488) as a one-line struct member access, so it's probably inlined anyway. Both the Boring and OpenSSL providers do it this ways.